src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,399 @@
+        /// void DoSomething(int i, Derived1 d) { }
+        /// void DoSomething(string s, Derived2 d) { }
+        /// void DoSomething(int i, Derived3 d) { }
+        /// 
+        /// Base b;
+        /// DoSomething(1, [||]b);
+        ///
+        /// *void DoSomething(string s, Derived2 d) { }* is not the perfect match candidate function for
+        /// *DoSomething(1, [||]b)* because int and string are not ancestor-descendant relationship. Thus,
+        /// Derived2 is not a potential conversion type.
+        /// 
+        /// arguments: The arguments of invocation expression
+        /// parameters: The parameters of function
+        /// targetArgument: The target argument that contains target node
+        /// targetParamIndex: Output the corresponding parameter index of the target arugment if function returns true
+        /// <returns>
+        /// True, if arguments and parameters match perfectly.
+        /// False, otherwise.
+        /// </returns>
+        // TODO: May need an API to replace this function,
+        // link: https://github.com/dotnet/roslyn/issues/42149
+        private static bool IsArgumentListAndParameterListPerfectMatch(SemanticModel semanticModel, SeparatedSyntaxList<ArgumentSyntax> arguments,
+            ImmutableArray<IParameterSymbol> parameters, ArgumentSyntax targetArgument, CancellationToken cancellationToken, out int targetParamIndex)
+        {
+            targetParamIndex = -1; // return invalid index if it is not a perfect match
+
+            var matchedTypes = new bool[parameters.Length]; // default value is false
+            var paramsMatchedByArray = false; // the parameter with keyword params can be either matched by an array type or a variable number of arguments
+            var inOrder = true; // assume the arguments are in order in default
+
+            for (var i = 0; i < arguments.Count; i++)
+            {
+                // Parameter index cannot out of its range, #arguments is larger than #parameter only if the last parameter with keyword params
+                var parameterIndex = Math.Min(i, parameters.Length - 1);
+
+                // If the argument has a name, get the corresponding parameter index
+                var nameSyntax = arguments[i].NameColon?.Name;
+                if (nameSyntax != null)
+                {
+                    var name = nameSyntax.ToString();
+                    var found = false;
+                    for (var j = 0; j < parameters.Length; j++)
+                    {
+                        if (name.Equals(parameters[j].Name))
+                        {
+                            // Check if the argument is in order with parameters.
+                            // If the argument breaks the order, the rest arguments of matched functions must have names
+                            if (i != j)
+                                inOrder = false;
+                            parameterIndex = j;
+                            found = true;
```

Comment:
this sort of bool flow control is very difficult to read.  instead, just extract a helper function to do this work.  it can then easily replace this line with `return true;` (and the later line with `return false;`).  The caller (i.e. this code) just checks that value and passes up the `false` value if that's what it receives.

Reply:
I am not sure I fully understand this.
Do you mean I can replace https://github.com/dotnet/roslyn/blob/43a850e206df6a3e8015bdd20d731cbb6c3cad08/src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs#L278-L293
by a helper function, and that function returns true if it finds the corresponding parameter?

Reply:
yup!